src/main/java/org/jboss/ejb/protocol/remote/RemotingEJBDiscoveryProvider.java

```diff
@@ -403,13 +418,36 @@ void connectAndDiscover(URI uri, String clusterEffective) {
                 return;
             }
             outstandingCount.getAndIncrement();
+
+            // keep a record of this URI if it has an associated cluster (EJBCLIENT-325)
+            if (clusterEffective != null) {
+                urisByCluster.computeIfAbsent(clusterEffective, ignored -> Collections.newSetFromMap(new ConcurrentHashMap<>())).add(uri);
+            }
+
             final IoFuture<ConnectionPeerIdentity> future = doPrivileged((PrivilegedAction<IoFuture<ConnectionPeerIdentity>>) () -> getConnectedIdentityUsingClusterEffective(endpoint, uri, "ejb", "jboss", authenticationContext, clusterEffective));
             onCancel(future::cancel);
             future.addNotifier(outerNotifier, uri);
         }
 
         void countDown() {
             if (outstandingCount.decrementAndGet() == 0) {
+
+                // before calculating the search result, update DNR to remove crashed last members of clusters (EJBCLIENT-325)
+                for (String cluster : urisByCluster.keySet()) {
+                    Set<URI> uris = urisByCluster.get(cluster);
+                      if (uris != null && uris.size() == 1) {
```

Comment:
An what in case of following scenario:
cluster X
1. node A added to cluster 
2. node B added to cluster
3. node A crashes
4. node B crashes

at point 4 (if I understand correctly) we should have 2 nodes for cluster X in urisByCluster map and also two entries in clusterFailedURI but. IMO nothing will happen and it should cause the last node in the cluster crashes.

Reply:
If there is a delay between node A crashing and node B crashing, node B will learn that node A has crashed (via a change to the client mappings membership) and it will send a topology update back to the client which does not include node A. This will cause the DNR to be updated to remove node A. If they crash at more or less the same time, what you say will hold.

Reply:
The failure I am seeing in the test case is more mystifying .... need to investigate.

Reply:
I'm not sure if I understand correctly here:
We add URI to urisByCluster each time connection attempt is made and never remove from it (or do we?).

"Node B will learn that node A has crashed (via a change to the client mappings membership)"

How this would affect urisByCluster?

Reply:
I've tried to write this answer three times and not getting anywhere :-)
1. The DNR is only updated by topology and module updates - no other way. The client is dependent on the outside world for learning about the outside world.
2. In order to "encourage" the outside world to tell more about what is going on, the client can try to establish connections to the outside world (i,e. nodes) and , if the client is lucky, these connection attempts might trigger topology updates - e.g. if a client has a configured connection to a node and that node happens to be a member of a cluster, the connection attempt will trigger a topology/module update and thereby update the DNR with more current information.
3. So, a discovery attempt is structured into two phases: a number of connection attempts (to refresh the DNR as well as keep track of failed destinations), followed by a direct search of the DNR to find all ServiceURLs which match a given filterspec. 
4. So, during the connection attempt phase, we keep track of which non-failed, cluster member URIs we know about are keep track of them. We also keep track of ConnectExceptions to know which nodes are really dead. This is not the same as failedDestinations which could have failed for security reasons. When all the connection attempts are completed and before we do the search, we know (i) which non-failed cluster members we know about according to the DNR (we build the connection attempt URIs from info in the DNR) and we know which URIs really are dead. 

So, the urisByCluster variable only exists for the lifetime of the connection attempt phase of this particular discovery call, is used to check the condition for removal of the cluster entry form the DNR (i.e. cluster contains a single node and that node is dead), so no need to maintain this variable over multiple calls.

Also, if Node B learns of the crash of Node A and manages to get a topology update back to the client, the DNR will be updated with that information (i.e. node A will be removed from the cluster in the DNR). Then on subsequent calls, in the connection phase, Node A will no longer appear as a known, non-failed node that we connect to and so not appear in urisByCluster.

You can think of urisByCluster as a very quick snapshot of the set of nodes we know about during a particular discovery attempt, for the purposes of making a quick calculation about whether or not  an entry in the DNR is stale.

Reply:
Sorry for being so long-winded!

Reply:
OK. Thanks for the exhaustive explanation. That is clear to me now.